solutions/usaco-916.mdx

```diff
@@ -0,0 +1,55 @@
+---
+id: usaco-916
+source: USACO Bronze 2019 February
+title: The Great Revegetation
+author: Ryan Chou
+---
+## Solution 1
+[Official Analysis (C++)](http://www.usaco.org/current/data/sol_revegetate_bronze_feb19.html)
+
+## Solution 2
+We can iterate through the favorite pastures, and increment one to the pasture with the greater index, making the result minimal. All pastures start with 
+the seed choice as one, this makes it so that we don't have to worry about decreasing or increasing more to a certain pasture. We'll also cycle through the 
+previous favorite pastures to make sure that our new alteration doesn't interfere with any of the other cow's favorite pastures.
+
+## Implementation - Solution 2
+**Time Complexity:** $\mathcal{O}(N^2)$
+
+<LanguageSection>
+<PySection>
+
+```py
+import sys
+sys.stdin = open('revegetate.in', 'r')
+sys.stdout = open('revegetate.out', 'w')
+
+pasture_num, cow_num = map(int, input().split())
+pastures = []
+ans = [1] * pasture_num
+
+for _ in range(cow_num):
+	past1, past2 = sorted(map(int,input().split()))
+	# subtract one so that the indexes start from 0
+	pastures.append([past1 - 1, past2 - 1])
+
+# assign the pastures at the front first.
+pastures.sort()
+for a, b in pastures:
+	# If they're not unique.
+	if ans[a] == ans[b]:
+		# Increment one to the pasture closer to the end.
+		# 100 is larger than 001.
+		largest_pos = max(a, b)
+		ans[largest_pos] += 1
+
+		# All favorite pastures up to the current one.
+		for i, j in pastures[:pastures.index([a, b])]:
+			# Increment again if requirement isn't satisfied.
+			if ans[i] == ans[j]:
+				largest_pos = max(a, b)
```

Comment:
shouldn't this be `max(i, j)` (but you sorted them already) or am i bad

actually i think the only time it will mess up previous pastures is when i or j equals b, since you're not changing any other part of the answer

Reply:
1. you want to increment the current two positions, if you increment the previous ones, you'll have to do another check.
2. you're checking the two indexes, not against a or b. I tried adding that in the code just to make sure, and it gets the sample test case wrong.

Reply:
wait so initially, none of ans[i] will equal ans[j]
but then you changed ans[max(a, b)]
so the only time when ans[i] will equal ans[j] is if i or j equals max(a, b)?

idk i might be wrong

Reply:
when we increment b in a, b, this works for the current favorite pastures only, so we have to go through the previous ones to make sure that b didn't change any of the others.
> none of ans[i] will equal ans[j]
we're assuming that they should, but if they aren't, we have to increment b again.
> so the only time when ans[i] will equal ans[j] is if i or j equals max(a, b)?
yea, but i feel like the current code is clearer. ig i could add it as a comment though.

Reply:
> to make sure that b didn't change any of the others.

yeah, this is what i mean- b is the only one that can affect the others

>>none of ans[i] will equal ans[j]

i meant that none of ans[i] will equal ans[j] before we increment ans[b]

>yea, but i feel like the current code is clearer. ig i could add it as a comment though.

sure, but up to you

Reply:
they could have a pasture in common with b, but this doesn't mean that i and j are the same. They have to be the same to have b increment.

Reply:
err yeah, i mean smth like
```
if i == b or j == b:
	if ans[i] == ans[j]:
```

Reply:
> err yeah, i mean smth like
> 
> ```
> if i == b or j == b:
> 	if ans[i] == ans[j]:
> ```

this works, but i feel like it's unnecessary, just ans[i] == ans[j] works. i'll add that as a comment